### DISCUSSION TITLE
Extending schedules to include temperature for use with Thermostats

### DISCUSSION THREAD

**Discussion by summongod** [Beormund/Tasmota32-Multi-Zone-Heating-Controller#2] [Ideas]
I testing various ways to control my heating system the requirements are:
1. Self contained - I don't want to use Home Assistant, NodeRed or similar as a backend. Simplicity = reliability
2. Simple commodity hardware -  I currently am soak testing and 8 channel relay board with a ESP32-C3-12F controller
I currently have this working with Xiaomi Mijia Bluetooth Thermometers using Tasmota BLE and some Tasmota Rules that extract temperature readings and then set Thermostat which controls the relays.

Is it possible to extend the scheduler to set target temperatures rather than just off and on?

I can see two ways this could be achieved:
1. Use TASMOTA THERMOSTATS and have the scheduler to THERMOSTAT commands setting the TEMPTARGETSET for each zone. (THERMOSTAT already exists and seems quite flexible and sophisticated)
2. Use SENSOR readings within the scheduler to determine whether the relay is activated. (More elegant, smaller footprint, less dependences)

I've been using THERMOSTAT for a few months on my test rig and it seems very good but I'm not sure that the level of sophistication is needed for most heating systems and so a simple zone target temperature (perhaps with a boost function) would be sufficient for most home systems.

**Comment by Beormund** [comment]
Would the schedule On state be ANDed with the temperature trigger? I.e. if scheduled time is ON and sensor within range THEN turn on relay?

**Reply by summongod** [reply]
Good question. I was thinking about sensors being thresholds rather than ranges.
Yes that sounds correct.

Are you thinking that temperature is part of the zone or the schedule? I suppose you would need to make two concurrent schedules with different temps if it was needed to have different temps in different zones

When thinking about temperatures I don't really think about explicit off settings as the same is achieved by setting a very low target temperature.

**Reply by Beormund** [reply]
Well zones have a 1:1 relationship with relays so it would probably be a matter of specifying a "set temp" threshold in the configure heating zone UI. That would be straightforward to add in the UI and persistence engine. The issue is how best to integrate the sensor trigger so it is loosely coupled and not hard coded to one input. Probably a matter of a callback that you can hook into and/or a custom command that you can call from MQTT/Upstream or via rules.

Internally the heating controller tracks the schedule or override on/off state for each zone. So if a set temp threshold is enabled and set then it's just a matter of adding an additional bit flag for thermostat state.

**Reply by summongod** [reply]
To drive the relays in thermostat I use.
Rule2
ON ATC2f8842#Temperature!="null" DO TEMPMEASUREDSET1 %value% ENDON
So I presume similar idea could work by setting the measured temp in the schedule or zone? Adding the linkage in autoexec so that sensors and zones can be easily changed etc?

**Reply by Beormund** [reply]
OK. Let me see what I can do in the coming weeks

**Reply by Beormund** [reply]
One issue having the thermostat combined with scheduler is the scheduler zone status shows ON when a schedule is active but the relay shows OFF when the room temp falls below set temp. Another issue is that if Alexa integration is enabled and switches off the relay the heating controller switches the zone from Auto to Advance Mode etc. So there are a few more complexities to think through. One way around this would be to unsubscribe the relay trigger when a temp threshold causes a relay state change and then subscribe the relay trigger afterwards. In a traditional heating system the zone controller/scheduler would have its own relays and each stat would have its own relay in series. Eg., If controller relay is on AND thermostat relay is on THEN call boiler for heat...

**Reply by summongod** [reply]
Yes, this is the same problem when having lights that switch off with inactivity. They have three states which need to be understood.
1/ Off
2/ On Off, when scheduled and inactive.
3/ On On, when scheduled and active. 

Would it make sense to have a scheduler that is more general purpose so it could be used for other things as well?
This would catch the Alexa issue as Alexa would control the schedule enable not the active/inactive.

**Reply by Beormund** [reply]
No UI yet but you can now set target temperature for a zone by using the command zone\<x\> {"temp": {"target": 22.4}} where \<x\> is zone. e.g., zone1. You can then send room temperature using zone\<x\>   {"temp": {"room": 22}}. If room temp is less than target temp for the zone the heating controller will send a low level GPIO command to switch the relay and bypass Tasmota's  relay handling code. If this works for you then we can add the UI elements for managing target temperature and display whether or not the zone is calling for heat. You should be able to send the zone command from within your rule. Let me know how you get on.

**Reply by Beormund** [reply]
If this approach works at a zone level it will be straightforward to extend it to schedules. It will then be possible to set differerent target temperatures for different times of the day.

**Reply by summongod** [reply]
Very cool, I'll play tomorrow when I have some time and update you.

**Reply by summongod** [reply]
Just testing quickly and I can't get the relay to fire. Works manually from the UI but if I set the room temp below the target nothing happens.

**Reply by Beormund** [reply]
And that's with a schedule running and powered on?

**Reply by summongod** [reply]
Interesting, just looked down and the relay had switched on. I've just raised the room temp to above the set temp and it's not gone off yet. How long should it take between a temp reading changing and the comparison made resulting in the relay being switched?

**Reply by Beormund** [reply]
Mmm- should be instant. Did the zone room temp command register in the console?

**Reply by summongod** [reply]
It looks like if the room temp changes then the comparison is made but not if the target temp is changed.

**Reply by Beormund** [reply]
Yep. I have fixed this. Will update shortly.

**Reply by Beormund** [reply]
OK. Pushed a quick fix for target temp trigger. I've started on schedule target temp. Should be available tomorrow. Thanks for carrying out the testing. Much appreciated.

**Reply by summongod** [reply]
No worries, the test rig (ESP32 plus patched leds) is left running so I can easily test. Once it is working I'll try to move it onto the real setup which is on a ESP32-C3-12F with BLE and a full relay set.

**Reply by Beormund** [reply]
Hit a bit of a snag with the low level gpio approach. Because of Interlock and other Tasmota features, all states of all relays are recalculated when one relay is changed which basically screws up temperature control of the gpio pins. So have had to pause and rethink.

**Reply by Beormund** [reply]
@summongod  - version 2 of heating controller released with thermostat support and Virtual Alexa Device support. Took a bit longer than expected as virtual Alexa device support only just implemented in Tasmota32 / Berry. Note, the architecture has changed quite a bit. There are now separate Tasmota Apps (tapps) for different functionality to minimise RAM.

**Reply by summongod** [reply]
Cool, I'm travelling for the next few days but I'll play on Friday. Any recommendation for the version of Tas should be using?

**Reply by Beormund** [reply]
The latest dev branch [tasmota32.bin](http://ota.tasmota.com/tasmota32/tasmota32.bin) or the lvgl version if you have a display/touchscreen.

**Reply by summongod** [reply]
I need BLE support so I'll try one of those versions first.

**Comment by GVollaro** [comment]
I have tested the thermostat with the room temperature setting and it works regularly.
If the zone is in an "ON" scheduling band, the actual status of the relay, on or off in relation to the room temperature < or => of the set threshold, is not available on the display which always gives the indication "ON"; therefore it would be desirable to also have an indication of the actual state of the relay.
A further improvement would be the possibility of setting a hysteresis value in order to avoid continuous ON / OFF when the room temperature varies around the set threshold value.

**Reply by Beormund** [reply]
Thanks for the feedback @GVollaro. By display I take it you mean the LCD display? So in addition to a schedule on/off status you'd like a heating/idle indication? Or maybe a flame 🔥 to indicate calling for heat for each zone (i.e. the relay state)? The original purpose of this project was simply to replicate a heating timer controller so I'm wondering if incorporating hysteresis is more a thermostat function?

**Comment by GVollaro** [comment]
I understand the initial purpose but with the introduction of the temperature the project has taken the lead of a chronothermostat and the introduction of the hysteresis is a further step towards the realization of a more sophisticated thermostat.
The idea of ​​indicating the actual state of the relay with a flame icon 🔥 is beautiful.
In another discussion started by me I launched the idea of ​​an evolution of the project towards an autonomous thermostat with an OpenTherm interface.
Currently I use Home Assistant and MQTT (graphical interface attached) but with the introduction in Tasmota of LVGL and Berry script language I would like to create something similar based only on Tasmota and which does not need HA and MQTT.
![Termostato_HA](https://user-images.githubusercontent.com/58565788/161719541-7105964b-5562-4c2f-b323-6a7c8cf9675a.JPG)
